### PR TITLE
drivers: digital-io: max22196 : Add support for MAX22194

### PR DESCRIPTION
**Pull request by RaduSabau1**
## Pull Request Description

MAX22194 is the same as MAX22196 except it doesn't have the 5 to 8 channels and channel counting functionality, therefore
the first commit handles the compatibility of the driver for MAX22194 as well as the MAX22196 using a chip_id enum inside of the device descriptor as well as the initialization parameter.

The second commit only uses the new approach for avoiding memory leaks at initialization of the driver.

The third commit updates the README.rst file so any MAX22194 extra related documentation is included and therefore be found in the no-OS repository.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

File: drivers/digital-io/max22196/iio_max22196.c [hunk 1]
```diff
@@ -835,12 +850,19 @@ int max22196_iio_init(struct max22196_iio_desc **iio_desc,
...
+
 	*iio_desc = descriptor;
 
 	return 0;
 
+free_dev:
+	max22196_remove(descriptor->max22196_desc);
 free_desc:
 	no_os_free(descriptor);
+
```
Comment (amiclaus) [inline drivers/digital-io/max22196/iio_max22196.c:863]: unrelated.

File: drivers/digital-io/max22196/iio_max22196.c [hunk 2]
```diff
@@ -301,20 +301,7 @@ static struct iio_attribute max22196_debug_attrs[] = {
 		.attributes = max22196_attrs	\
 	}
 
-static struct iio_channel max22196_channels[] = {
-	MAX22196_CHANNEL(0),
-	MAX22196_CHANNEL(1),
-	MAX22196_CHANNEL(2),
-	MAX22196_CHANNEL(3),
```
Comment (amiclaus) [inline drivers/digital-io/max22196/iio_max22196.c:308]: is this change related to max22194 support?

Comment (RaduSabau1) [reply drivers/digital-io/max22196/iio_max22196.c:308]: Yes, channels in the iio driver will now be allocated dynamic according to the chip id

**Comment by RaduSabau1** [comment]
V2

- Delete unrelated changes

**Review by amiclaus — COMMENTED** [review]
please perform a rebase with latest main branch to make sure all check are passing.

File: drivers/digital-io/max22196/iio_max22196.c [hunk 3]
```diff
@@ -783,6 +770,34 @@ static int max22196_iio_write_global_cfg(void *dev, char *buf, uint32_t len,
...
+ * physical channels.
+ * @param desc - The iio device descriptor.
+ * @return 0 in case of success, an error code otherwise.
+ */
+int max22196_iio_setup_channels(struct max22196_iio_desc *desc)
+{
+	struct iio_channel *max22196_iio_channels;
+	int i, num_ch;
+
+	max22196_iio_channels = no_os_calloc(desc->max22196_desc->chip_id ? 2 : 1,
```
Comment (rbolboac) [inline drivers/digital-io/max22196/iio_max22196.c:784]: this memory should be freed in max22196_iio_remove

File: drivers/digital-io/max22196/max22196.c [hunk 4]
```diff
@@ -198,6 +198,9 @@ int max22196_set_mode(struct max22196_desc *desc, uint32_t ch,
 	if (ch > MAX22196_CHANNELS - 1)
 		return -EINVAL;
 
+	if (desc->chip_id == ID_MAX22194 && ch > MAX22194_CHANNELS - 1)
```
Comment (rbolboac) [inline drivers/digital-io/max22196/max22196.c:201]: all in all the code would look much nicer if you would have in desc the maximum channels numbers which you add at init based on device_id... then you simply check ch against that number...

**Comment by RaduSabau1** [comment]
V3

- Fixed memory leak.
- Handling the MAX22194 channel count at remove as well.
- Addressed other comments.

**Comment by RaduSabau1** [comment]
V4

- Added max_chn_nb variable in the device descriptor structure.

File: drivers/digital-io/max22196/iio_max22196.c [hunk 5]
```diff
@@ -783,6 +770,34 @@ static int max22196_iio_write_global_cfg(void *dev, char *buf, uint32_t len,
...
+	struct iio_channel *max22196_iio_channels;
+	int i, num_ch;
+
+	max22196_iio_channels = no_os_calloc(desc->max_chn_nb,
+					     sizeof(*max22196_iio_channels));
+
+	if (!max22196_iio_channels)
+		return -ENOMEM;
+
+	num_ch = desc->max_chn_nb;
```
Comment (rbolboac) [inline drivers/digital-io/max22196/iio_max22196.c:790]: is this variable needed? you should be able to use max_chn_nb instead

File: drivers/digital-io/max22196/max22196.c [hunk 6]
```diff
@@ -239,7 +239,7 @@ int max22196_filter_set(struct max22196_desc *desc, uint32_t ch,
 {
 	int ret;
 
-	if (ch > MAX22196_CHANNELS - 1)
+	if (ch > desc->max_chn_nb)
```
Comment (rbolboac) [inline drivers/digital-io/max22196/max22196.c:242]: please use the same check in all APIs where chn is used

File: drivers/digital-io/max22196/max22196.c [hunk 7]
```diff
@@ -466,7 +472,10 @@ int max22196_get_chan_cnt(struct max22196_desc *desc, uint32_t ch,
 	int ret;
 	uint32_t reg_val;
 
-	if (ch > MAX22196_CHANNELS - 1)
+	if (desc->chip_id == ID_MAX22194)
+		return -EINVAL;
+
+	if (ch > desc->max_chn_nb)
```
Comment (rbolboac) [inline drivers/digital-io/max22196/max22196.c:478]: can be merged in same if condition

**Comment by RaduSabau1** [comment]
V5

- Compacted code where possible
- Fixed wrong if condition
- IIO channels structure is removed by the iio_remove() function.

File: drivers/digital-io/max22196/max22196.c [hunk 8]
```diff
@@ -272,6 +272,9 @@ int max22196_filter_get(struct max22196_desc *desc, uint32_t ch,
 	int ret;
 	uint32_t del_val;
 
+	if (desc->chip_id == ID_MAX22194 && ch > MAX22194_CHANNELS - 1)
+		return -EINVAL;
+
 	if (ch > MAX22196_CHANNELS - 1)
 		return -EINVAL;
```
Comment (rbolboac) [inline drivers/digital-io/max22196/max22196.c:279]: can't this be replaced with if (ch > desc->max_chn_nb - 1) ?

File: drivers/digital-io/max22196/max22196.c [hunk 9]
```diff
@@ -508,6 +511,7 @@ int max22196_init(struct max22196_desc **desc,
 	descriptor = no_os_calloc(1, sizeof(*descriptor));
 	if (!descriptor)
 		return -ENOMEM;
+
```
Comment (rbolboac) [inline drivers/digital-io/max22196/max22196.c:514]: unrelated change to this commit

File: drivers/digital-io/max22196/max22196.c [hunk 10]
```diff
@@ -560,10 +567,12 @@ int max22196_remove(struct max22196_desc *desc)
...
-		ret = max22196_set_chan_cnt(desc, i, MAX22196_CHN_CNT_RESET);
-		if (ret)
-			return ret;
+	if (desc->comm_desc) {
+		for (i = 0; i < desc->max_chn_nb; i++) {
+			ret = max22196_set_chan_cnt(desc, i, MAX22196_CHN_CNT_RESET);
+			if (ret)
+				return ret;
+		}
 	}
```
Comment (rbolboac) [inline drivers/digital-io/max22196/max22196.c:573]: please do not forget to free max22196_iio_channels, it should be done in this commit because here is where you are allocating the memory.

File: drivers/digital-io/max22196/iio_max22196.c [hunk 11]
```diff
@@ -873,6 +871,7 @@ int max22196_iio_remove(struct max22196_iio_desc *iio_desc)
 	if (!iio_desc)
 		return -ENODEV;
 
+	no_os_free(iio_desc->iio_dev->channels);
```
Comment (rbolboac) [inline drivers/digital-io/max22196/iio_max22196.c:874]: as mentioned, this should be moved in the previous commit

**Comment by RaduSabau1** [comment]
V6

- Moved related changes to their commits.
- Addressed other comments.

**Review by amiclaus — DISMISSED** [review]

**Comment by RaduSabau1** [comment]
V7

- Using a more formal approach in README.rst

File: drivers/digital-io/max22196/max22196.c [hunk 12]
```diff
@@ -422,7 +422,7 @@ int max22196_set_chan_cnt(struct max22196_desc *desc, uint32_t ch,
 	int ret;
 	uint8_t msb_byte, lsb_byte;
 
-	if (ch > MAX22196_CHANNELS - 1)
+	if (desc->chip_id == ID_MAX22194 || ch > desc->max_chn_nb - 1)
```
Comment (CiprianRegus) [inline drivers/digital-io/max22196/max22196.c:425]: What should this evaluate to? I would rewrite the expression as it's a bit hard to follow.

Comment (RaduSabau1) [reply drivers/digital-io/max22196/max22196.c]: Will try them splitting into two if conditions, since they are not related one with the other.

**Review by rbolboac — APPROVED** [review]
